text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,182 @@
+
+`pack create-builder sclevine/builder -b builder.toml`
+
+builder.toml:
+```toml
+[stack]
+id = "io.buildpacks.stacks.bionic"
+build-image = "example.com/build"
+run-image = "example.com/run"
+run-image-mirrors = ["example.org/run"]
+
+[extend]
+run-image = "sclevine/run"
+run-image-mirrors = ["scl.sh/run"]
+[extend.run] # run-metadata-toml-file (stored on new run image metadata)
+packages = ["git"]
+[extend.build] # build-metadata-toml-file (stored on builder image metadata)
+packages = ["git"]
+```
+
+Behavior: creates a new builder with additional packages as well as a new run image (`sclevine/run`) with additional packages.
+
+If `extend.run-image` is not specified, `extend.run` is stored on the builder metadata and used to dynamically extend the run image on `pack build`.
+
+### Extended an Existing Builder
+
+`pack create-builder sclevine/builder -b builder.toml`
+
+builder.toml:
+```toml
+[builder]
+image = "example.com/builder"
+
+[extend]
+run-image = "sclevine/run"
+run-image-mirrors = ["scl.sh/run"]
+[extend.run] # run-metadata-toml-file (stored on new run image metadata)
+packages = ["git"]
+[extend.build] # build-metadata-toml-file (stored on new builder image metadata)
+packages = ["git"]
+```
+
+Behavior: creates a new version of an existing builder with additional packages as well as a new run image (`sclevine/run`) with additional packages.
+
+If `extend.run-image` is not specified, `extend.run` is stored on the new builder metadata and used to dynamically extend the run image on `pack build`.
+
+Note that the `[builder]` section is mutually exclusive with `[stack]`.
+
+### Building an App with Additional Packages
+
+`pack build sclevine/myapp`
+
+project.toml:
+```toml
+...
+[run.extend] # run-metadata-toml-file (stored on new run image metadata)
+packages = ["git"]
+[build.extend] # build-metadata-toml-file (stored on new builder image metadata)
+packages = ["git"]
+...
+```
+
+Behavior: creates a version of the current builder with additional packages and an ephemeral run image with additional packages, then does a normal `pack build`.
+
+Question: should we store the new builder image to make rebuild faster? If so, where? Should we generate a tag for it?
+
+### Upgrade an Extended Builder
+
+Both build and run images: `pack upgrade sclevine/builder`
+
+Run images (including mirrors): `pack upgrade sclevine/run`
+
+### Upgrade an App
+
+`pack upgrade sclevine/myapp`
+
+This will run `pack upgrade sclevine/run`, generate an ephemeral image, and then do the equivalent of `pack rebase sclevine/myapp`.
+
+Attempting to rebase an app directly after it's been upgraded is not permitted and will fail, because the current base is now ephemeral.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Specification
+
+Platforms may extend app images using the above interface by running the `extend` executable as root in a new container and creating an image from the result.
+Extending an image should generate a single layer.
```

Comment:
Is this solely a side-effect given the fact that extend is executed once and how it would be implemented?

Couldn't a build system execute `extend` potentially be able to yield multiple layers (for instance if it wants to leverage it's own caching)?

Reply:
It could, but it would probably need to look like multiple `extend` scripts (as the slices interface doesn't work when you already have layers). Do you have a suggestion for an interface that would result in multiple layers (considering that the generated layers depend on arbitrary input)?